### PR TITLE
PR: Implement autosave for the Editor

### LINKED ISSUE #2111: Enhancement: Implement autosave of editor files every X minutes
_From [googlelogin@bjoern-kahl.de](https://code.google.com/u/googlelogin@bjoern-kahl.de/) on 2015-01-12T16:59:36Z_

Feature request / Enhancement:
1) Currently, when Spyder becomes unresponsive or just crashes, one loses all changes in edited files since last manual save.
2) Spyder does save files before a "run" action, but since other things than running the code can crash Spyder, that is not enough protection against data loss. 

What is the expected output? What do you see instead?

 1) Spyder should automatically save a file X minutes after the last manual or automatic save. 
2) Spyder should offer a preference setting which allows the user to specify the time interval X between automatic saves and also allows to disable the automatic save.
3) The automatic save should go into a special directory "backups" under the ~/.spyder directory (or any other user-configurable place, including the parent directory of the edited file itself) and take the edited file's name with a prefix and/or suffix attached.  It must not use the original file for auto-saving. What version of the product are you using? On what operating system? Spyder 2.3.2, Mac OSX 10.6

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=2111_

### LINKED ISSUE #4409: Provide autosave functionality to not lose unsaved code when Spyder crashes
## Description
I don't know how to reproduce, but this has happened for the 3rd time in a few weeks:
1. I launch Spyder through Anaconda 1.4.3
2. Spyder loads with my usual 3..5 small .py files opened
3. I start writing and executing code in one of them, using the Editor, IPython console, Variable explorer and Help panes.
4. Sometimes Spyder crashes. It happens not during code execution or anything, but instead happens when I'm not doing anything. It's just sitting there and boom. And of course all code that I've written after the last execution is lost.

## Version and main components

* Spyder Version: 3.1.2 
* Python Version: 3.6.0
* Qt Versions: 5.6.2, PyQt5 5.6 on Windows

## Dependencies
```
pyflakes >=0.6.0 :  1.5.0 (OK)
pep8 >=0.6       :  1.7.0 (OK)
pygments >=2.0   :  2.1.3 (OK)
qtconsole >=4.2.0:  4.2.1 (OK)
nbconvert >=4.0  :  4.2.0 (OK)
pandas >=0.13.1  :  0.19.2 (OK)
numpy >=1.7      :  1.11.3 (OK)
sphinx >=0.6.6   :  1.5.1 (OK)
rope >=0.9.4     :  0.9.4-1 (OK)
jedi >=0.8.1     :  0.9.0 (OK)
matplotlib >=1.0 :  2.0.0 (OK)
sympy >=0.7.3    :  1.0 (OK)
pylint >=0.25    :  1.6.4 (OK)

```

### PR DESCRIPTION
**Pull request by jitseniesen**
### Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 3.x)
* [x] Followed [PEP8](https://www.python.org/dev/peps/pep-0008/) for code style
* [x] Ensured your pull request hasn't eliminated unrelated blank lines/spaces,
      modified the ``spyder/defaults`` directory, or added new icons/assets
* [x] Wrote at least one-line docstrings for any new functions
* [x] Added at least one unit test covering the changes, if at all possible
* [x] Described your changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed
* [x] Included a screenshot, if this PR makes any visible changes to the UI

### To do
* [x] Editor reloading file because it changed on disk should not trigger an autosave

Following CAM Gerlach's review:
* [x] Shorten text at top of recovery dialog and use tooltips on buttons instead
* [x] Reformat list of files in recovery dialog, maybe using table widget or rules
* [x] Try making `Size:` and `Last modified:` italics
* [x] Use `localtime` and `'%Y-%m-%d %H:%M:%S'` format when listing time
* [x] Add preferences for turning autosave off and specifying autosave interval
* [x] Format docstring of `recovery_data_key_function`
* [x] Format docstrings of test functions to comply with relevant PEP standard
* [x] Include a test for a graceful handling of when the autosave is unsuccessful
* [x] Further tweak to text in recovery dialog window

## Description of Changes

Every minute, save a copy of every file in the editor with unsaved changes in a new autosave directory, which lives under the spyder config directory. This autosave file is removed when it is no longer needed (e.g. when the user saves the file manually). On startup, check the autosave directory and offer to recover any files which are found there. This should enable the user to recover most of their work in the case of a crash.

I will write a more complete description in the comment below.

### Issue(s) Resolved

Fixes #2111

### CODE REVIEW

**Comment by jitseniesen** [comment]
This is the promised more detailed description.

Every minute, the editor goes through all opened files. If any file has unsaved changes, then the contents of the editor are saved in an autosave file. The autosave file has the same name as the original file, but it stored in the `autosave` directory under the Spyder configuration directory. If the autosave file would have the same name as an existing autosave file, then a suffix is added to the same. For instance, if the user opens two files with the name `foo.py` in different directories and edits both files, then one of the autosave files will be called `foo.py` and the other `foo-1.py`. Files created by "New file" which have not yet been saved by the users (these files have names like `untitled0.py`) are not autosaved. Spyder keeps track of a dict mapping original file names to autosave file names and stores this as `autosave_mapping` in the Spyder configuration file.

When the user saves a file, then the autosave file will be deleted, because it is not needed anymore. When the user closes a file and does not save the changes made, then the autosave file will also be deleted. Thus, when Spyder is quit cleanly, all autosave files will be deleted.

When Spyder starts up, it checks the autosave directory. If it contains any files, for instance because the previous invocation of Spyder crashed, then the user is given the option to recover them. This is done by a modal dialog window like this:

recover

The design is admittedly not very exciting, but I don't have much experience with design UI elements and it does the job. I am happy to change it if you have suggestions.

The names of the original files are taken from the `autosave_mapping` config variable. If this does not work (because the config file did not save properly, is corrupted, or deleted by the user), then the original file is listed as "not recorded".

For every autosave file, the user has three options. The `Recover` button replaces the original file by the autosave file. The `Discard` button removes the autosave file and leaves the original file unchanged. The `Open` button opens the original file (if it is known and it exists) and the autosave file in the editor after the Recover dialog window closes, so that the user can investigate the situation and decide on the best action. After one of the buttons is clicked, the row of buttons is deactivated (greyed out). When all rows are deactivated and thus the user has taken some action on all files, the window closes. The `Cancel` button closes the window without taking any further action.

Implementationwise, I had to change only two existing classes. Unfortunately, these are the most bloated classes in our code base: `Editor` (the plugin) and `EditorStack` (the widget). In an effort to isolate the autosave functionality, I added a classes called `AutosaveComponent` with composition to both `Editor` and `EditorStack`. The recovery functionality is implemented in a new class called `RecoveryDialog`.

One potential controversial thing in my implementation is that I added a field called `changed_since_autosave`, for keeping track of whether a file was edited after the last autosave, to the `QDocument` of the editor. It feels naughty to add a field to an external class, but I don't see a disadvantage with it, and the advantage is that different editors with the same file share the `QDocument`, so they also share the `changed_since_autosave` flag.

**Comment by ccordoba12** [comment]
This looks really, really nice!! Thanks a lot @jitseniesen! The failures seems unrelated to your work.

I'll try to take a look at your work this weekend because right now I'm busy trying to release 3.3.1 and beta1.

File: spyder/widgets/recover.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+                autosave_dict = gather_file_data(full_name)
+                self.data.append((orig_dict, autosave_dict))
+        except FileNotFoundError:  # autosave dir does not exist
+            pass
+        self.data.sort(key=recovery_data_key_function)
+        self.num_enabled = len(self.data)
+
+    def add_label(self):
+        """Add label with explanation at top of dialog window."""
+        txt = _('Spyder found some autosave files. You can use the autosave '
```
Comment (CAM-Gerlach) [inline spyder/widgets/recover.py]: Much too wordy and somewhat redundant; also "recover" doesn't actually match the wording of the button ("restore"). I suggest you revise it to simply read something like ``Autosave files found. What would you like to do?\n\nThis will be shown each startup if autosave files are not restored, moved or deleted.`` and then add the following tooltips to each button:

Restore: ``Recover the autosaved file to its original location, replacing the original if it exists.``
Discard: ``Delete the autosaved file.``
Open: ``Display the autosave (and the original, if it exists) in Spyder's Editor, and move or delete it manually.``

**Comment by jitseniesen** [comment]
Thanks @CAM-Gerlach for your comments. I agree with all of them and will implement them, though I will probably wait until the split-plugins PR is merged. Ideally, @ccordoba12 will have at least a look through my description of the PR, if not the details in the code, in order to spot any big changes which may be needed. I'd also like to know what he thinks about the configuration options (turn autosave on/off; change autosave interval); I agree that they are very useful but we want to be careful with adding new options.

The backup system is a nice idea for extension, but I definitely consider it out of scope for this PR. I'd like some user feedback first.

**Comment by CAM-Gerlach** [comment]
>  I'd also like to know what he thinks about the configuration options (turn autosave on/off; change autosave interval); I agree that they are very useful but we want to be careful with adding new options.

Right, although considering essentially every other similar program (editor, word processor, etc) that I know of that has an autosave feature has at least these two options, I'd find it hard to justify not including them.

> The backup system is a nice idea for extension, but I definitely consider it out of scope for this PR. I'd like some user feedback first.

Yeah, for sure—it's well out of scope here.

Also, on another note, did you include a test for a graceful handling of when the autosave is unsuccessful (e.g. can't write to the disk, permissions issue, random I/O issue)? I didn't see one listed but I might have missed it.

**Comment by CAM-Gerlach** [comment]
@jitseniesen Thanks!

A few quick comments before I take a nap—the separate table cells for the Restore/Discard/Open Buttons look somewhat awkward. At the very least, can combine them into one cell? You can also title the column something appropriate, like ``Actions``. Finally, it would look visually much more readable to make the original/autosave file table cells left-aligned, since you are displaying several fields (Size, Date Modified).

**Comment by ccordoba12** [comment]
I'm trying to fix some errors reported here in PR #7908.

**Comment by jitseniesen** [comment]
I got rid of the vertical grid lines. New screen shot:
recover

**Comment by CAM-Gerlach** [comment]
That looks really clean, @jitseniesen ; much better!

One minor tweak (to my own suggested text, I believe) on the second line:

> This *dialog* will be shown *again on next* startup if *any* autosave files are not restored, moved or deleted.

**Comment by ccordoba12** [comment]
@jitseniesen, the error reported by Travis seems legit. Please take a look at it.

**Comment by jitseniesen** [comment]
> the error reported by Travis seems legit. Please take a look at it.

I agree. Unfortunately I can't reproduce it and the cause is not clear to me, so I have to guess a bit. Since the error is reported during teardown, it may be caused by the autosave timer firing while or after the plugin is closed, so I added some code to prevent this. It's not a bad idea anyway.

**Comment by chrisfeigl** [comment]
Thank you to everyone who is involved in building this. Keep it up, please! :)

File: spyder/plugins/editor/plugin.py [hunk 2]
```diff
@@ -2950,3 +3002,72 @@ def set_create_new_file_if_empty(self, value):
         """Change the value of create_new_file_if_empty"""
         for editorstack in self.editorstacks:
             editorstack.create_new_file_if_empty = value
+
+
+class AutosaveComponent(object):
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py]: Let's move this class to `plugins/editor/utils/autosave.py` because

a) It's pure Python.
b) We don't need to add more content to the (already very big) Editor plugin file.

File: spyder/plugins/editor/widgets/editor.py [hunk 3]
```diff
@@ -2516,6 +2541,117 @@ def dropEvent(self, event):
         event.acceptProposedAction()
 
 
+class AutosaveComponent:
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/editor.py]: * Please also move this class to `plugins/editor/utils/autosave.py`.
* Also, please make this inherit from `object` to make it a new class in Python 2.

**Comment by jitseniesen** [comment]
Screenshot with new autosave config options at the bottom:

autosave-config

**Comment by jitseniesen** [comment]
> I left a couple of minor comments and I'll do a more thorough (and hopefully final) review this week.

Great! This is indeed ready as far as I am concerned (assuming the CI tests pass).

I like your suggestion to move the autosave components to their own files, so I did that.

**Comment by CAM-Gerlach** [comment]
@jitseniesen I can do some functional testing too, hopefully tomorrow.

I haven't gone over the code in detail nor yet tested this behavior, but given we've advertised this feature as the solution to the reports we've been getting of Spyder completely wiping users' files on occasion (including multiple times to our own @jnsebgosselin ), I wanted to cross-post the relevant parts of the discussion @jnsebgosselin and I had a couple weeks ago on Gitter about how autosave should work, in order to "fail safe" and properly mitigate such issues (one of, if not *the* primary motivation to add it currently), to make sure that's included here:

@jnsebgosselin said:

> I'm wondering about something regarding the auto-save functionality @jitseniesen is implementing.
This is also related to what people are saying in spyder-ide/spyder#2111 and spyder-ide/spyder#4409 . I'm saying this because on my new laptop, I have a small SSD and a large secondary HDD. I have a problem currently with the secondary HDD. From time to time, it simply disconnect. Du to this, it happened a couple of times that Spyder would wipe entirely the files that were opened when trying to save them or running them. Not only the unsaved changes were lost, but the entire content of the files were gone. I'm wondering if some of the people reporting in the aforementioned issues are experiencing that also.

@CAM-Gerlach said:

> I've seen a number of similar reports, that we typically dismissed as isolated incidents, user error or not Spyder related at the time. That's the real value of the autosave functionality, not just for the unsaved changes but having an extra, separate backup copy that gets touched separately to the main one and, to provide the best protection against this issue, is persistent during a session (not being overwritten/removed after a user-triggered save, only on exit if Spyder exits cleanly with everything successfully saved) and not being directly overwritten during each autosave, but rather an (e.g. timestamped) new version is saved, and the old one only deleted until after confirming the new autosave was fully successful, with perhaps a few second delay to ensure everything is flushed to disk).

@jnsebgosselin said:

> OK, that's what I was wondering. So it is more like an "autosave and recovery" feature than just a simple and dumb autosave functionality. If that's the case, we should change the title of the PR to take that into account because I think this is really important.

@CAM-Gerlach said:

> I have no idea if that's exactly how it works. That's my naive suggestion of how it should work to start, to minimize the chance of data loss like you and others have described. The recovery UI is pretty neat though, and is an important part of it. A more advanced version of the feature, not really in scope for that PR, would save multiple versions of the document on a timer or every time it was saved, like most other editors and IDEs (NP++, Sublime, Pycharm, Eclipse, etc) offer.

@jnsebgosselin said:

> Ah ok thanks! I agree with you on this. Anyway, I have a mean to reproduce the total wiping of file content until I send my laptop for repair if needed

@CAM-Gerlach said:

> Yeah, the versioning would be nice. But just making the basic implementation of the autosave itself work like I described (not overwriting/removing the autosave when the file is saved by the user, only on a clean Spyder exist, and writing to a new file each time and only on confirmed success deleting the old, or warning the user that it failed) should (if I understand the failure mode correctly) ensure at least one good copy exists for the great majority of these events other that specific corner cases or user mistakes.
>
> To really be safe against the file blanking problem, you would also want to include a check before deletion of an old autosave file and after writing the new one (or main copy, for when the autosave is deleted on a clean Spyder exit or manually closing a file in the Editor) that reads back the new/main one and refuses to delete the old/autosave and warns the user if either any error occurred during the test read of the new/main, or if the new/main save has >= ~3 bytes of content while the old one has e.g. > ~100 B; unless the user wiped an entire file of more than a few lines and left the empty file for some reason without deleting it entirely this would properly prevent meaningful data loss due to the file blanking problem and many other causes with few to no false positives; otherwise, the autosave feature might not protect against at least certain incarnations of the problem it was supposed to stop in cases where the error occurred silently, only to the main file or after the last autosave, was accidentally overwritten or corrupted, was due to some external hardware or software interaction with the file after save, etc.

**Comment by jitseniesen** [comment]
@CAM-Gerlach It's an autosave, it's not a backup. It is inspired by the autosave system in Emacs, though I think most autosave systems work similarly (or they save periodically to the actual file opened without user interaction, which I think would be against what our users expect).

Work can be lost in all kind of ways and autosave does not protect against all failure modes. From the reports, the main failure mode (apart from user error) is that Spyder / the OS / the computer crashes and the user loses unsaved files. This should help against it.

It may not help against hard disk errors. Specifically, if Spyder saves the file and it thinks this is done succesfully, then the autosave file is deleted. So this autosave feature does not help if due to a hardware failure the file did not really save but Spyder thinks it did save. This may be what @jnsebgosselin experienced: Spyder think it saved the file, but in fact the file got wiped. I am not sure how often this happens, it seems pretty rare (I don't remember any report apart from his), but I think the solution would be hardening the save function, perhaps add an fsync or try reading back the file or I don't know.

Keeping the autosave file after the file is saved makes everything more complicated. I'm not sure it's worth the effort, and in any case I think it's best left to another iteration.

**Comment by ccordoba12** [comment]
@jitseniesen, thanks for the quick response. Could you move `AutosaveComponentForEditorPlugin` and `AutosaveComponentForEditorStack` to `plugins/editor/utils/autosave.py` and rename them `AutosaveComponentForPlugin` and `AutosaveComponentForStack`, or even better, `AutosaveForPlugin` and `AutosaveForStack`?

My suggestion has to do with the fact that I don't think they need to be put next to the plugin and in `editor/widgets` because these classes don't correspond to gui components, they are utilites to perform the autosave functionality.

**Comment by ccordoba12** [comment]
> Keeping the autosave file after the file is saved makes everything more complicated

I don't like this either because it pollutes your filesystem with backup files. I also agree with @jitseniesen that we need to leave that for another PR and it needs to be opt-in, not opt-out.

**Comment by CAM-Gerlach** [comment]
> It's an autosave, it's not a backup.

Which technically *is* a backup, just a particular type of one (and not a substitute for a traditional one), intended specifically to protect against data loss or corruption due to application/OS hangs, crashes, power failure, and other such instances outside the user's control. From that viewpoint, crashes that wipe the data in user's open file(s) are intimately part of that narrow scope (as opposed to e.g. versioning, persistent backup, or other such protection against hardware failure or user error, etc).

> I am not sure how often this happens, it seems pretty rare (I don't remember any report apart from his)

Aside from the multiple times he experienced it, there were at least two others, #6567 and #6571 , and I'm pretty sure I saw at least one or two more in another thread. In both of those, the error occurred when running a file, before which all files are automatically saved by default (putting them all potentially at risk), and if the fatal crash occurred anytime after the ``FileInfo._write_to_file()`` call returns, then the autosave will fail to protect against this. Furthermore, it also isn't any help in any situation where the crash/bad write occurs when on or around a save where an autosave didn't exist (the file was previously saved less than 60 seconds ago, or if the file was only modified in the 60 seconds before save).

Although this is certainly rarer than just loosing unsaved work, the consequences are vastly more severe (total data loss of one or more entire files) and for a non-VCS managed project (or one that is not committed frequently), and the same good user habits (saving frequently) that insure against the former type of data loss not only give no protection for this, but actually make it more likely (more saves, and less likely for autosave to kick in.

> I think the solution would be hardening the save function

Yes, we should certainly implement that as well, but it doesn't cover the window where the file is saved after Spyder has deleted the autosave and not regenerated one. Aside from not deleting the autosave copy after saves and only when Spyder is closed cleanly, a solution would be to have the save() function itself trigger an autosave before it executed if none existed already, do the save with flush and fsync, and then delete it after if no errors were detected (perhaps with a user-optional readback check). 

> I don't like this either because it pollutes your filesystem with backup files.

That's not really true for @jitseniesen 's implementation, since autosave files all go in the ``autosave`` directory under the hidden Spyder config directory, not scattered throughout the filesystem. Furthermore, they are still deleted on a clean exit and the only case where they are retained under our proposal relative to currently is if a file is saved, and then Spyder crashes sometime within 60 seconds (or before the user makes more changes), which is generally when a file wipe would be most likely to occur anyway.

> I think it's best left to another iteration.

> we need to leave that for another PR

Okay, sure, I guess. Perhaps @jnsebgosselin could do it if he has the time, since he's the one who can actually test if it works.

> it needs to be opt-in, not opt-out

Perhaps readback verification could be opt-in, but the changes necessary (more robust saves as described above, and/or not removing the autosave files after save until clean exit) to protect against Spyder completely wiping users' files on save should be at least opt-out, since as a rare but potentially catastrophic problem users are unlikely to know to enable it until its too late, and I haven't seen any reports of this happening in other major IDEs.

**Comment by CAM-Gerlach** [comment]
If @jnsebgosselin agrees and @ccordoba12 thinks this is getting too off topic, I can open a new dedicated issue and move this there.

**Comment by jnsebgosselin** [comment]
I support @CAM-Gerlach view on this. Being able to protect against data corruption (backup & restore) seems like a very important feature to have. Even though it is not a frequent issue, when this happens, the consequences are  absolutely disastrous.

But I also agree this should be discussed in another Issue.

**Comment by CAM-Gerlach** [comment]
> But I also agree this should be discussed in another Issue.

Okay, thanks. I've summarized everything in #8237 , and moved my responses over there as well (removing them here to reduce clutter).

**Comment by jitseniesen** [comment]
> Could you move `AutosaveComponentForEditorPlugin` and `AutosaveComponentForEditorStack` to `plugins/editor/utils/autosave.py` and rename them `AutosaveComponentForPlugin` and `AutosaveComponentForStack`, or even better, `AutosaveForPlugin` and `AutosaveForStack`?

Done. I misunderstood you before, but I now get that when you said these are pure Python, you meant they are not Qt and so should not be in the widgets dir.

**Comment by jitseniesen** [comment]
The failing test (`spyder/plugins/editor/lsp/tests/test_client.py::test_initialization`) fails locally on my computer also on master bc6d741d5d85087b8a6d7d068e7fc91773879fce ("Merge from 3.x: Release 3.3.2"), so I assume it is unrelated to this PR.

**Comment by jitseniesen** [comment]
I still don't know what happened with CI, but it seems to have corrected itself.

File: spyder/plugins/editor/widgets/tests/test_recover.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    # ham.py: Both original and autosave files exist, mentioned in mapping
+    assert osp.join(orig_dir, 'ham.py') in text(0, 0)
+    assert osp.join(autosave_dir, 'ham.py') in text(0, 1)
+
+    # spam.py: Only autosave file exists, mentioned in mapping
+    assert osp.join(orig_dir, 'spam.py') in text(1, 0)
+    assert 'no longer exists' in text(1, 0)
+    assert osp.join(autosave_dir, 'spam.py') in text(1, 1)
+
+    # eggs.py: Only original files exists, so cannot be recovered
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_recover.py]: Is this test missing?

File: spyder/plugins/editor/tests/test_plugin.py [hunk 5]
```diff
@@ -163,6 +176,73 @@ def test_no_template(setup_editor):
     shutil.move(osp.join(osp.dirname(template), 'template.py.old'), template)
 
 
+def test_editor_has_autosave_component(setup_editor):
+    """Test that Editor includes an AutosaveForPlugin."""
+    editor, qtbot = setup_editor
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/plugins/editor/tests/test_plugin.py]: In fact, its supposed to not even work at all in Pytest 4 and I was going to do a pass through all our tests to remove it before Pytest 4 landed, but I didn't get to it in time.

File: spyder/plugins/editor/utils/autosave.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        Remove autosave file for specified file.
+
+        This function also updates `self.autosave_mapping` and clears the
+        `changed_since_autosave` flag.
+        """
+        filename = fileinfo.filename
+        if filename not in self.name_mapping:
+            return
+        autosave_filename = self.name_mapping[filename]
+        os.remove(autosave_filename)
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/plugins/editor/utils/autosave.py]: Of course, we *do* want to still report an error when the autosave itself fails, so users know they don't have their safety net (and something could be going on with their disk, which could make normal saves fail).

File: spyder/plugins/editor/widgets/recover.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+                autosave_dict = gather_file_data(full_name)
+                self.data.append((orig_dict, autosave_dict))
+        except FileNotFoundError:  # autosave dir does not exist
+            pass
+        self.data.sort(key=recovery_data_key_function)
+        self.num_enabled = len(self.data)
+
+    def add_label(self):
+        """Add label with explanation at top of dialog window."""
+        txt = _('Autosave files found. What would you like to do?\n\n'
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/plugins/editor/widgets/recover.py]: Yeah, if you want, although both will actually work—this gets into some fine-grained syntax distinctions that even I don't want to think about, jeje. An "Autosaved file" is a file that has been autosaved, i.e. using "autosave" as a past tense verb, while an "Autosave file" is a file generated by the Autosave system, i.e. using Autosave, the proper noun that's the name for the system as a (proper) adjective. The former is slightly clearer, I guess, so we can go with that throughout. Regardless, the bigger deal is just to pick one and be consistent throughout.

File: spyder/plugins/editor/widgets/tests/test_editor.py [hunk 8]
```diff
@@ -574,5 +576,157 @@ def test_tab_copies_find_to_replace(editor_find_replace_bot, qtbot):
...
+
+def test_autosave_ignores_error(editor_bot, mocker):
+    """Test that autosave() ignores errors when writing to file."""
+    editor_stack, editor = editor_bot
+    mock_write = mocker.patch.object(editor_stack, '_write_to_file')
+    try:
+        mock_write.side_effect = PermissionError
+    except NameError:  # Python 2
+        mock_write.side_effect = IOError
+    editor_stack.autosave.autosave(0)
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/plugins/editor/widgets/tests/test_editor.py]: Pytest will automatically fail the test if it generates an exception, if that's what you're referring to,

Furthermore, IMO we should not simply hide the exception but give the user a friendly warning dialog that the autosave failed, so they're aware they no longer have safety net is no longer in place and potentially give them warning of an I/O problem with the drive, as their actual main-file saves may also fail especially if they are to the same drives.

File: spyder/plugins/editor/utils/autosave.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/editor/utils/autosave.py]: > Of course, we do want to still report an error when the autosave itself fails

Sure, but this is probably not the best place to do it.

File: spyder/plugins/editor/widgets/recover.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/editor/widgets/recover.py]: Ok, let's leave `Autosave` then.

File: spyder/plugins/editor/widgets/tests/test_editor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/editor/widgets/tests/test_editor.py]: @jitseniesen?

**Comment by ccordoba12** [comment]
@jitseniesen, this needs a merge with master.

**Comment by CAM-Gerlach** [comment]
@jitseniesen This is almost done and it looks like it only needs a merge and a few minor tweaks. Looking forward to seeing this in Spyder!

File: spyder/plugins/editor/tests/test_plugin.py [hunk 9]
```diff
@@ -13,23 +13,34 @@
...
+    return mock
+
+
 @pytest.fixture
 def setup_editor(qtbot, monkeypatch):
     """Set up the Editor plugin."""
-    qapplication()
-    from spyder.plugins.editor.plugin import Editor
-
+    # monkeypatch.setattr('spyder.dependencies', Mock())
```
[the remark this answers is not in the document]
Comment (jitseniesen) [reply spyder/plugins/editor/tests/test_plugin.py]: ok

File: spyder/plugins/editor/tests/test_plugin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jitseniesen) [reply spyder/plugins/editor/tests/test_plugin.py]: This has been fixed in the meantime by commit ba113f430a1

File: spyder/plugins/editor/utils/autosave.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jitseniesen) [reply spyder/plugins/editor/utils/autosave.py]: Errors when removing autosave files are now ignored. Regarding errors when autosave itself fails, see the discussion of ` spyder/plugins/editor/widgets/tests/test_editor.py ` line 702 below.

File: spyder/plugins/editor/widgets/tests/test_editor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jitseniesen) [reply spyder/plugins/editor/widgets/tests/test_editor.py]: I was not sure about the dialog, because if the failure is persistent then a dialog would be displayed every minute, which is rather annoying. But now that autosave can be turned off, I agree that we should inform the user if the autosave fails.

[the remark this answers is not in the document]
Comment (jitseniesen) [reply spyder/plugins/editor/widgets/tests/test_editor.py]: I like that.

[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/editor/widgets/tests/test_editor.py]: I agree with @CAM-Gerlach's suggestion too.

[the remark this answers is not in the document]
Comment (jitseniesen) [reply spyder/plugins/editor/widgets/tests/test_editor.py]: Done.

File: spyder/plugins/editor/widgets/tests/test_recover.py (hunk 4 above)
Comment (jitseniesen) [reply spyder/plugins/editor/widgets/tests/test_recover.py]: > Is this test missing?

No, but I added a comment to clarify.

File: spyder/plugins/editor/utils/autosave.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jitseniesen) [reply spyder/plugins/editor/utils/autosave.py]: I changed my mind with the new error reporting mechanism and now I think it's best tot be explicit so errors when removing autosave files are reported.

**Comment by jitseniesen** [comment]
Assuming all tests pass, this is ready for another round of review.

**Comment by CAM-Gerlach** [comment]
Thanks so much @jitseniesen for all your amazing work! I'll do a (hopefully final) review and test soon.

**Comment by scratchmex** [comment]
I just lost a day of work because Windows restarted to update and I haven't saved the file. In any case, the files created with New File and named `untitled0.py` should be autosaved if they are open in the editor as opposed to the comment by @jitseniesen:

>
>Files created by "New file" which have not yet been saved by the users (these files have names like `untitled0.py`) are not autosaved. Spyder keeps track of a dict mapping original file names to autosave file names and stores this as `autosave_mapping` in the Spyder configuration file.
> 

It would be a useful feature.
